## Instructions {-}

1. You may talk to a friend, discuss the questions and potential directions for solving them. However, you need to write your own solutions and code separately, and not as a group activity. 

2. Do not write your name on the assignment.

3. Write your code in the *Code* cells and your answer in the *Markdown* cells of the Jupyter notebook. Ensure that the solution is written neatly enough to understand and grade.

4. Use [Quarto](https://quarto.org/docs/output-formats/html-basics.html) to print the *.ipynb* file as HTML. You will need to open the command prompt, navigate to the directory containing the file, and use the command: `quarto render filename.ipynb --to html`. Submit the HTML file.

5. The assignment is worth 100 points, and is due on **6th October 2022 at 11:59 pm**. 

# I
USA's GDP per capita from 1960 to 2021 is given by the tuple `T` in the code cell below. The values are arranged in ascending order of the year, i.e., the first value is for 1960, the second value is for 1961, and so on.

In [7]:
T = (3007, 3067, 3244, 3375,3574, 3828, 4146, 4336, 4696, 5032,5234,5609,6094,6726,7226,7801,8592,9453,10565,11674,12575,13976,14434,15544,17121,18237,19071,20039,21417,22857,23889,24342,25419,26387,27695,28691,29968,31459,32854,34515,36330,37134,37998,39490,41725,44123,46302,48050,48570,47195,48651,50066,51784,53291,55124,56763,57867,59915,62805,65095,63028,69288)

## I-1.
### I-1(a)
Use list comprehension to produce a list of the gaps between consecutive entries in `T`, i.e, the increase in GDP per capita with respect to the previous year. The list with gaps should look like: [60, 177, ...].

*(6 points)*

In [8]:
GDP_gaps = [T[i+1] - T[i] for i in range(len(T) - 1)]

In [9]:
GDP_gaps

[60,
 177,
 131,
 199,
 254,
 318,
 190,
 360,
 336,
 202,
 375,
 485,
 632,
 500,
 575,
 791,
 861,
 1112,
 1109,
 901,
 1401,
 458,
 1110,
 1577,
 1116,
 834,
 968,
 1378,
 1440,
 1032,
 453,
 1077,
 968,
 1308,
 996,
 1277,
 1491,
 1395,
 1661,
 1815,
 804,
 864,
 1492,
 2235,
 2398,
 2179,
 1748,
 520,
 -1375,
 1456,
 1415,
 1718,
 1507,
 1833,
 1639,
 1104,
 2048,
 2890,
 2290,
 -2067,
 6260]

### I-1(b)
Use the list comprehension developed in (a) to find the maximum gap size, i.e, the maximum increase in GDP per capita.

*(2 points)*

In [10]:
max(GDP_gaps)

6260

The maximum gap gize is 6260

### I-1(c)
Use the list comprehension developed in (a) to find the percentage of gaps higher than $1000.

*(5 points)*

In [11]:
len([x for x in GDP_gaps if x > 1000]) / len(GDP_gaps)

0.5245901639344263

Approximately 52.459% of gaps are higher than $1000

## I-2 
### I-2(a)
Create a dictionary `D`, where the `key` is the year, and `value` for the `key` is the increase in GDP per capita in that year with respect to the previous year, i.e., the gaps computed in part (1). 

*(6 points)*

In [12]:
D = {}

for year in range(1961, 2022):
    D[year] = GDP_gaps[year - 1961]
    
D

{1961: 60,
 1962: 177,
 1963: 131,
 1964: 199,
 1965: 254,
 1966: 318,
 1967: 190,
 1968: 360,
 1969: 336,
 1970: 202,
 1971: 375,
 1972: 485,
 1973: 632,
 1974: 500,
 1975: 575,
 1976: 791,
 1977: 861,
 1978: 1112,
 1979: 1109,
 1980: 901,
 1981: 1401,
 1982: 458,
 1983: 1110,
 1984: 1577,
 1985: 1116,
 1986: 834,
 1987: 968,
 1988: 1378,
 1989: 1440,
 1990: 1032,
 1991: 453,
 1992: 1077,
 1993: 968,
 1994: 1308,
 1995: 996,
 1996: 1277,
 1997: 1491,
 1998: 1395,
 1999: 1661,
 2000: 1815,
 2001: 804,
 2002: 864,
 2003: 1492,
 2004: 2235,
 2005: 2398,
 2006: 2179,
 2007: 1748,
 2008: 520,
 2009: -1375,
 2010: 1456,
 2011: 1415,
 2012: 1718,
 2013: 1507,
 2014: 1833,
 2015: 1639,
 2016: 1104,
 2017: 2048,
 2018: 2890,
 2019: 2290,
 2020: -2067,
 2021: 6260}

### I-2(b)
Use the dictionary `D` to find the year when the GDP per capita increase from the previous year was the maximum.

*(4 points)*

In [13]:
[key for key, value in D.items() if D[key]==max(D.values())]

[2021]

The GDP per capita increase from the previous year was the maximum in 2021.

### I-2(c)
Use the dictionary `D` to find the years when the GDP per capita decreased with respect to the previous year.

*(4 points)*

In [14]:
[key for key, value in D.items() if D[key] < 0]

[2009, 2020]

The GDP per capita decreased in 2009 and 2020.

# II
## II-1
Read the [data](https://raw.githubusercontent.com/cwkenwaysun/TEDmap/master/data/TED_Talks.json) on ted talks from 2006 to 2017.

*(2 points)*

In [15]:
import pandas as pd

tedTalks = pd.read_json('https://raw.githubusercontent.com/cwkenwaysun/TEDmap/master/data/TED_Talks.json')

tedTalks

id           speaker  \
0        7       David Pogue   
1        6      Craig Venter   
2        4        Burt Rutan   
3        3      Ashraf Ghani   
4        5      Chris Bangle   
...    ...               ...   
2469  2740    Carina Morillo   
2470  2797   Justin Davidson   
2471  2801       Anne Lamott   
2472  2799       Tim Ferriss   
2473  2804  Richard Browning   

                                               headline  \
0                                      Simplicity sells   
1                              Sampling the ocean's DNA   
2                  The real future of space exploration   
3                         How to rebuild a broken state   
4                              Great cars are great art   
...                                                 ...   
2469              To understand autism, don't look away   
2470  Why glass towers are bad for city life -- and ...   
2471          12 truths I learned from life and writing   
2472  Why you should define your fears instead of yo...   
2473                             How I built a jet suit   

                                        URL  \
0        http://www.ted.com/talks/view/id/7   
1        http://www.ted.com/talks/view/id/6   
2        http://www.ted.com/talks/view/id/4   
3        http://www.ted.com/talks/view/id/3   
4        http://www.ted.com/talks/view/id/5   
...                                     ...   
2469  http://www.ted.com/talks/view/id/2740   
2470  http://www.ted.com/talks/view/id/2797   
2471  http://www.ted.com/talks/view/id/2801   
2472  http://www.ted.com/talks/view/id/2799   
2473  http://www.ted.com/talks/view/id/2804   

                                            description  \
0     New York Times columnist David Pogue takes aim...   
1     Genomics pioneer Craig Venter takes a break fr...   
2     In this passionate talk, legendary spacecraft ...   
3     Ashraf Ghani's passionate and powerful 10-minu...   
4     American designer Chris Bangle explains his ph...   
...                                                 ...   
2469  Carina Morillo knew almost nothing about autis...   
2470  There's a creepy transformation taking over ou...   
2471  A few days before she turned 61, writer Anne L...   
2472  The hard choices  what we most fear doing, ask...   
2473  We've all dreamed of flying  but for Richard B...   

                                         transcript_URL  month_filmed  \
0     http://www.ted.com/talks/view/id/7/transcript?...             2   
1     http://www.ted.com/talks/view/id/6/transcript?...             7   
2     http://www.ted.com/talks/view/id/4/transcript?...             2   
3     http://www.ted.com/talks/view/id/3/transcript?...             7   
4     http://www.ted.com/talks/view/id/5/transcript?...             2   
...                                                 ...           ...   
2469  http://www.ted.com/talks/view/id/2740/transcri...            10   
2470  http://www.ted.com/talks/view/id/2797/transcri...             3   
2471  http://www.ted.com/talks/view/id/2801/transcri...             4   
2472  http://www.ted.com/talks/view/id/2799/transcri...             4   
2473  http://www.ted.com/talks/view/id/2804/transcri...             4   

      year_filmed             event duration date_published  \
0            2006           TED2006  0:21:26        6/27/06   
1            2005    TEDGlobal 2005  0:16:51     2004/05/07   
2            2006           TED2006  0:19:37       10/25/06   
3            2005    TEDGlobal 2005  0:18:45       10/18/06   
4            2002           TED2002  0:20:04     2004/05/07   
...           ...               ...      ...            ...   
2469         2016  TEDxRiodelaPlata  0:07:09      2006/7/17   
2470         2017            TEDNYC  0:12:39      2006/6/17   
2471         2017           TED2017  0:15:55      2006/9/17   
2472         2017           TED2017  0:13:21     2006/12/17   
2473         2017           TED2017  0:07:08        6/13/17   

      

## II-2
Find the number of talks in the dataset.

*(2 points)*

In [16]:
tedTalks.shape[0]

2474

There are 2474 talks in the dataset.

## II-3 
Find the `headline`, `speaker` and `year` of the talk with the highest number of views.

*(5 points)*

In [17]:
position_max_views = tedTalks['views'].argmax()
tedTalks.iloc[position_max_views,:]

print("The talk with the highest number of views...")
print("headline: ", tedTalks.iloc[position_max_views,2])
print("speaker: ", tedTalks.iloc[position_max_views, 1])
print("year_filmed: ", tedTalks.iloc[position_max_views, 7])

The talk with the highest number of views...
headline:  Do schools kill creativity?
speaker:  Ken Robinson
year_filmed:  2006


##  II-4 
Do the majority of talks have less views than the average number of views for a talk? Justify your answer.

*(4 points)*

**Hint:** Print summary statistics for questions (4) and (5).

In [18]:
tedTalks.describe()

id  month_filmed  year_filmed         views
count  2474.000000   2474.000000  2474.000000  2.474000e+03
mean   1433.782943      5.587712  2011.299515  1.747851e+06
std     797.970133      3.443934     3.665506  2.584138e+06
min       1.000000      1.000000  1972.000000  5.065300e+04
25%     768.250000      2.000000  2009.000000  7.703258e+05
50%    1451.500000      5.000000  2012.000000  1.151252e+06
75%    2118.750000      9.000000  2014.000000  1.748603e+06
max    2804.000000     12.000000  2017.000000  4.792402e+07

The mean number of views is higher than the number of views at 50% so, yes, the majority of talks have less views than the average number of views.

## II-5 
Do at least 25% of the talks have more views than the average number of views for a talk? Justify your answer.

*(4 points)*

Yes, at least 25% of talks have more views than the average number of views for a talk. This can be seen in the summary statistics where the number of views at 75% is 1.748603e+06 which is higher than the mean of 1.747851e+06 views. This shows that, at the very least, the views of the talks in the 75th to 100th percentile (which is a total of 25% of the talks: 100 - 75 = 25) must have more views than the average.  

## II-6 

### II-6(a)
The last column of the dataset consists of votes obtained by the talk under different categories, such as *Funny, Confusing, Fascinating, etc.* For each category, create a new column in the dataset that contains the votes obtained by the tedtalk in that category. Print the first 5 rows of the updated dataset.

*(20 points)*

In [19]:
for i in range(tedTalks.shape[0]):
    for category in tedTalks.loc[i, 'rates']:
        tedTalks.loc[i, category['name']] = category['count']
        
tedTalks.head()

id       speaker                              headline  \
0   7   David Pogue                      Simplicity sells   
1   6  Craig Venter              Sampling the ocean's DNA   
2   4    Burt Rutan  The real future of space exploration   
3   3  Ashraf Ghani         How to rebuild a broken state   
4   5  Chris Bangle              Great cars are great art   

                                  URL  \
0  http://www.ted.com/talks/view/id/7   
1  http://www.ted.com/talks/view/id/6   
2  http://www.ted.com/talks/view/id/4   
3  http://www.ted.com/talks/view/id/3   
4  http://www.ted.com/talks/view/id/5   

                                         description  \
0  New York Times columnist David Pogue takes aim...   
1  Genomics pioneer Craig Venter takes a break fr...   
2  In this passionate talk, legendary spacecraft ...   
3  Ashraf Ghani's passionate and powerful 10-minu...   
4  American designer Chris Bangle explains his ph...   

                                      transcript_URL  month_filmed  \
0  http://www.ted.com/talks/view/id/7/transcript?...             2   
1  http://www.ted.com/talks/view/id/6/transcript?...             7   
2  http://www.ted.com/talks/view/id/4/transcript?...             2   
3  http://www.ted.com/talks/view/id/3/transcript?...             7   
4  http://www.ted.com/talks/view/id/5/transcript?...             2   

   year_filmed           event duration  ... Unconvincing Longwinded  \
0         2006         TED2006  0:21:26  ...        104.0       78.0   
1         2005  TEDGlobal 2005  0:16:51  ...          3.0        8.0   
2         2006         TED2006  0:19:37  ...         77.0       70.0   
3         2005  TEDGlobal 2005  0:18:45  ...         35.0       21.0   
4         2002         TED2002  0:20:04  ...         81.0      100.0   

  Informative Inspiring  Fascinating Confusing     OK  Persuasive  \
0       399.0     230.0        166.0      27.0  146.0       230.0   
1       199.0      73.0        163.0      23.0   10.0        23.0   
2       337.0     501.0        247.0      23.0  115.0       202.0   
3       218.0     114.0         44.0      25.0   51.0       112.0   
4        88.0     163.0         76.0      32.0   64.0        28.0   

   Jaw-dropping  Obnoxious  
0          54.0      142.0  
1          74.0        0.0  
2          63.0       47.0  
3          35.0       11.0  
4           9.0       45.0  

[5 rows x 30 columns]

### II-6(b)
With the data created in (a), find the `headline` of the talk that received the highest number of *Confusing* votes.

*(5 points)*

In [20]:
position_max_confusing = tedTalks['Confusing'].argmax()
print("headline of most confusing TedTalk: ", tedTalks.iloc[position_max_confusing,2])

headline of most confusing TedTalk:  I believe we evolved from aquatic apes


### II-6(c)
With the data created in (a), find the `headline` and the `year` of the talk that received the highest percentage of votes in the *Fascinating* category. 

$$\text{Percentage of } \textit{Fascinating} \text{ votes for a ted talk} = \frac{Number \ of \  votes \ in \ the \ category - \ Fascinating}{Total \ votes \ in \ all  \ categories}$$

*(10 points)*

In [21]:
for i in range(tedTalks.shape[0]):
    total_votes = 0
    for category in tedTalks.loc[i, 'rates']:
        total_votes += category['count']
    tedTalks.loc[i, 'Fascinating Percentage'] = tedTalks.loc[i, 'Fascinating'] / total_votes
        
position_max_fas_per = tedTalks['Fascinating Percentage'].argmax()
tedTalks.iloc[position_max_fas_per,:]

print("The talk that received the highest percentage of votes in the Fascinating category was... ")
print("headline: ", tedTalks.iloc[position_max_fas_per,2])
print("year: ", tedTalks.iloc[position_max_fas_per, 7])

The talk that received the highest percentage of votes in the Fascinating category was... 
headline:  Could a Saturn moon harbor life?
year:  2009


# III

## III-1
Download the data set “univ.txt”. Read it with python.

*(2 points)*

In [22]:
univ_txt = pd.read_csv('univ.txt',sep='\t')
univ_txt

Name             Location   Rank  \
0                        Princeton University        Princeton, NJ    1.0   
1                          Harvard University        Cambridge, MA    2.0   
2                       University of Chicago          Chicago, IL    3.0   
3                             Yale University        New Haven, CT    3.0   
4                         Columbia University         New York, NY    5.0   
..                                        ...                  ...    ...   
226    University of Massachusetts--Dartmouth  North Dartmouth, MA  220.0   
227         University of Missouri--St. Louis        St. Louis, MO  220.0   
228  University of North Carolina--Greensboro       Greensboro, NC  220.0   
229        University of Southern Mississippi      Hattiesburg, MS  220.0   
230                     Utah State University            Logan, UT  220.0   

     Tuition and fees  Undergrad Enrollment  
0             45320.0                5402.0  
1             47074.0                6699.0  
2             52491.0                5844.0  
3             49480.0                5532.0  
4             55056.0                6102.0  
..                ...                   ...  
226           19270.0                7295.0  
227           26277.0               13569.0  
228           21595.0               15951.0  
229           16094.0               11840.0  
230           19772.0               25259.0  

[231 rows x 5 columns]

## III-2 
### III-2(a) 
Find summary statistics of the data. Based on the statistics, answer parts b-e.

*(1 point)*

In [23]:
univ_txt.describe()

Rank  Tuition and fees  Undergrad Enrollment
count  231.000000        231.000000            231.000000
mean   113.982684      33769.246753          14946.619048
std     65.995518      10756.733516          10569.664095
min      1.000000       5300.000000           1001.000000
25%     56.000000      25693.000000           6238.500000
50%    111.000000      31608.000000          12949.000000
75%    171.000000      42721.000000          22145.500000
max    220.000000      55056.000000          54513.000000

### III-2(b) 
How many universities are there in the data set?

*(2 points)*

In [24]:
univ_txt.shape[0]

231

There are 231 universities in the data set.

### III-2(c) 
Estimate the maximum tuition among universities that are in the bottom 25% when ranked by total tuition and fees.

*(3 points)*

According to the summary statistics, the maximum tuition among universities that are in the bottom 25% is $25,693 (when ranked by total tuition and fees).

### III-2(d) 
How many universities share the ranking of 220? (If `s` universities share the same rank, say `r`, then the next lower rank is `r+s`, and all the ranks in between `r` and `r+s` are dropped)

*(5 points)*

In [25]:
rank_220 = univ_txt.loc[:, 'Rank'] == 220.0

len(univ_txt.loc[rank_220,:])

12

12 universities share the ranking of 220.

### III-2(e)
Can you find the mean fee for an undergrad student in the US from the summary statistics? Justify your answer.

*(3 points)*

No, you cannot find the mean fee for an undergrad student in the US from the summary statistics. This is because the mean "Tuition and Fees" in the summary statistics is calculating the mean fee per university and not per undergrad student. Every university has a different number of students so the mean per undergrad student is not the same as mean per university.

## III-3	
Find the average `Tuition and fees` for an undergrad student in the US.

*(5 points)*

In [26]:
univ_txt["Total Tuition"] = univ_txt["Tuition and fees"] * univ_txt["Undergrad Enrollment"]

In [27]:
univ_txt["Total Tuition"]

total_tuition = 0
total_students = 0

for n in univ_txt["Total Tuition"]:
    total_tuition += n
    
for n in univ_txt["Undergrad Enrollment"]:
    total_students += n
    
print(total_tuition / total_students)

30845.32903298868


The average tuition and fees for an undergrad student in the US is $30,845.33